### PR TITLE
Filter examples

### LINKED ISSUE #133: Demo codeView font too small
Please increase the size of the text in the code view. It's so small that I have a very hard time reading it.

Linux Mint 17.1, pyqt, Python 3.4.2 (Anaconda)

Thanks

### PR DESCRIPTION
**Pull request by ntjess**
Users who know what they want, but don't know how to find it in the ever-growing list of examples can take solace in a lovely filter edit.

- [x] Fixes #133

Also optionally searches content

Updated video with all current capabilities:

https://user-images.githubusercontent.com/23620506/123526529-84735c00-d6a6-11eb-83de-bc515c8fab6f.mp4

### CODE REVIEW

**Comment by ixjlyons** [comment]
Cool feature, I'm :+1:

It could be useful to search the example text as well, such as to find uses of `ImageItem` where "image" may not be in the name. There you'd probably only want to connect to `editingFinished`. Maybe it could be enabled with a checkbox.

In any case, LGTM.

**Comment by ntjess** [comment]
@ixjlyons Great idea, something like this? It turns out the files can cache pretty easily and still respond to textChanged. I am fine with changing that if desired, though.

I decided to ignore case sensitivity, but am open to enforcing it.

https://user-images.githubusercontent.com/23620506/123500809-b1743000-d60e-11eb-93e3-b7e9c040e881.mp4

**Comment by j9ac9k** [comment]
First thing first, this is _slick_ ...thanks for this @ntjess !

Couple of issues, it doesn't run on PyQt6 due to the enum namespace.  

```
Traceback (most recent call last):
  File "/Users/ogi/.zinit/plugins/pyenv---pyenv/versions/3.9.1/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/ogi/.zinit/plugins/pyenv---pyenv/versions/3.9.1/Python.framework/Versions/3.9/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/ogi/Developer/pyqtgraph/examples/__main__.py", line 16, in <module>
    run()
  File "/Users/ogi/Developer/pyqtgraph/examples/ExampleApp.py", line 497, in main
    loader = ExampleLoader()
  File "/Users/ogi/Developer/pyqtgraph/examples/ExampleApp.py", line 309, in __init__
    self.codeLayout.addItem(QtWidgets.QSpacerItem(100,100,QtWidgets.QSizePolicy.Expanding,QtWidgets.QSizePolicy.Expanding), 0, 0)
AttributeError: type object 'QSizePolicy' has no attribute 'Expanding'
```

Should be `QSizePolicy.Policy.Expanding`. Everything other than PyQt6 still supports the Qt5 enum namespace, so this issue is only noticeable when running PyQt6.

Also, not sure if its feasible to do something about it, but the placeholder text doesn't look quite right in dark mode on macOS:

image

Looking at the diff here, it doesn't look like you hard-code any specific colors, so this just may be an upstream Qt issue; and I would say its probably not worth the effort to try and work-around it.

**Comment by ntjess** [comment]
@j9ac9k interesting... Can you run pyqt6's uic generator on the ui template? That's all I did; I want to see if it improperly uses enums for you as well.

Regarding the dark theme, is this something that can be detected using `app.property('darkMode')`?

File: examples/ExampleApp.py [hunk 1]
```diff
@@ -249,36 +256,78 @@ def match_multiline(self, text, delimiter, in_state, style):
...
-        self.codeBtn = QtGui.QPushButton('Run Edited Code')
-        self.codeLayout = QtGui.QGridLayout()
+        self.codeBtn = QtWidgets.QPushButton('Run Edited Code')
+        self.codeLayout = QtWidgets.QGridLayout()
         self.ui.codeView.setLayout(self.codeLayout)
         self.hl = PythonHighlighter(self.ui.codeView.document())
-        app = QtGui.QApplication.instance()
+        app = QtWidgets.QApplication.instance()
         app.paletteChanged.connect(self.updateTheme)
-        self.codeLayout.addItem(QtGui.QSpacerItem(100,100,QtGui.QSizePolicy.Policy.Expanding,QtGui.QSizePolicy.Policy.Expanding), 0, 0)
```
Comment (j9ac9k) [inline examples/ExampleApp.py:278]: This is the change that's causing the enum issue.

**Comment by j9ac9k** [comment]
> @j9ac9k interesting... Can you run pyqt6's uic generator on the ui template? That's all I did; I want to see if it improperly uses enums for you as well.

I ran the UIC on the template, but the issue was in ExampleApp.py, I put an inline comment that should point out the relevant change. 

> Regarding the dark theme, is this something that can be detected using `app.property('darkMode')`?

That is indeed how you can recognize if the app is in dark mode or not, but I _really_ don't like going away from the defaults for cosmetic stuff like this.  I put a comment there before looking at the diff to see if you specified a color, but on closer reflection you didn't do that, so I'm good with leaving it as the default behavior, even if its not ideal.

File: examples/ExampleApp.py (hunk 1 above)
Comment (ntjess) [reply examples/ExampleApp.py:278]: Right, is that the same output you get when you run your uic generator?

**Comment by ntjess** [comment]
Thanks. I didn't recall touching that line so I was a little confused.

I did specify a color (`#000`) for foreground text during a highlight, I'll have to change that before you merge

**Comment by j9ac9k** [comment]
I know it's tough to troubleshoot dark mode on Windows, let me know if I can help.

**Comment by ntjess** [comment]
@j9ac9k I tried unsuccessfully to test dark mode on my computer; are you able to test against the recent commits? Thanks

**Comment by NilsNemitz** [comment]
@ntjess , I like the filter 👍 

Another option for the text highlight would be to just set the color to something from the QPalette.

QtGui.QPalette.ColorRole.WindowText
and
QtGui.QPalette.ColorRole.Text
are the main foreground colors I guess.

Then you would get whatever Qt thinks the foreground color should be and (maybe) don't need to even detect the mode.

I think this makes a reasonable dark palette on Windows, if you want to mess around with it:
``` python
### Qt color definitions for dark palette on Windows #####################
def make_dark_QPalette(self):
    """ manually define a dark mode palette """
    BLACK      = QtGui.QColor('#000000')
    BG_LIGHT   = QtGui.QColor('#505354')
    BG_NORMAL  = QtGui.QColor('#2e3132')
    BG_DARK    = QtGui.QColor('#0e1112')
    FG_LIGHT   = QtGui.QColor('#f0f4f5')
    FG_NORMAL  = QtGui.QColor('#d4d8d9')
    FG_DARK    = QtGui.QColor('#b8bcbd')
    SEL_LIGHT  = QtGui.QColor('#148CD2')
    SEL_NORMAL = QtGui.QColor('#1464A0')
    SEL_DARK   = QtGui.QColor('#14506E')
    qpal = QtGui.QPalette( QtGui.QColor(BG_DARK) )
    for ptype in (  QtGui.QPalette.ColorGroup.Active,  QtGui.QPalette.ColorGroup.Inactive ):
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Window, BG_NORMAL )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.WindowText, FG_LIGHT ) # or white?
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Base, BG_DARK )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Text, FG_LIGHT )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.AlternateBase, BG_DARK )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.ToolTipBase, BG_LIGHT )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.ToolTipText, FG_LIGHT )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Button, BG_NORMAL )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.ButtonText, FG_LIGHT )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Link, SEL_NORMAL )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.LinkVisited, FG_NORMAL )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.Highlight, SEL_LIGHT )
        qpal.setColor( ptype, QtGui.QPalette.ColorRole.HighlightedText, BLACK )
    qpal.setColor( QtGui.QPalette.ColorGroup.Disabled, QtGui.QPalette.ColorRole.Button, BG_NORMAL )
    qpal.setColor( QtGui.QPalette.ColorGroup.Disabled, QtGui.QPalette.ColorRole.ButtonText, FG_DARK )
    qpal.setColor( QtGui.QPalette.ColorGroup.Disabled, QtGui.QPalette.ColorRole.WindowText, FG_DARK )
    return qpal

app = QtWidgets.QApplication.instance()
app.setPalette( dark_palette )
```

**Comment by ntjess** [comment]
@NilsNemitz thanks for the suggestion :+1: I am less interested in constructing an entire dark palette than just making sure highlighted text has the right foreground. That being said, the palette-based coloring seems more appropriate for at least determining the right text color

**Comment by j9ac9k** [comment]
> That being said, the palette-based coloring seems more appropriate for at least determining the right text color

This is absolutely the way to do it.

File: examples/exampleLoaderTemplate_pyqt5.py [hunk 2]
```diff
@@ -1,10 +1,11 @@
 # -*- coding: utf-8 -*-
 
-# Form implementation generated from reading ui file 'exampleLoaderTemplate.ui'
+# Form implementation generated from reading ui file 'examples/exampleLoaderTemplate.ui'
 #
-# Created by: PyQt5 UI code generator 5.12.3
+# Created by: PyQt5 UI code generator 5.15.4
```
Comment (j9ac9k) [inline examples/exampleLoaderTemplate_pyqt5.py:5]: Hmm...this _might_ be an issue, I'll investigate if this is backward compatible.

**Comment by ntjess** [comment]
@NilsNemitz @j9ac9k Looks like, at least for highlighting, the color discussion is irrelevant. Black text always shows up best against the yellow-highlighted background, so I can hard-code that

File: examples/ExampleApp.py [hunk 3]
```diff
@@ -5,8 +5,6 @@
 import subprocess
 from argparse import Namespace
 
-import qdarkstyle
```
Comment (ntjess) [inline examples/ExampleApp.py]: :man_facepalming: :man_facepalming: :man_facepalming: :man_facepalming:

Comment (j9ac9k) [reply examples/ExampleApp.py]: > Black text always shows up best against the yellow-highlighted background, so I can hard-code that

There is nothing hard-coded in Qt about the highlight color (unless you hardcode it yourself).  Generally speaking, the fewer hard specifications on colors for this sort of thing you can do, the better (aka let Qt sort out the OS integration bits).

**Comment by j9ac9k** [comment]
> @NilsNemitz @j9ac9k Looks like, at least for highlighting, the color discussion is irrelevant. Black text always shows up best against the yellow-highlighted background, so I can hard-code that

There is nothing hard-coded in Qt about the highlight color (unless you hardcode it yourself). Generally speaking, the fewer hard specifications on colors for this sort of thing you can do, the better (aka let Qt sort out the OS integration bits).

**Comment by j9ac9k** [comment]
@ntjess this LGTM; would you mind updating the original post to say "Fixes #133"?

**Comment by ntjess** [comment]
@j9ac9k Done!

**Comment by j9ac9k** [comment]
The zoom in/out/reset is really slick!  Thanks for making it happen.

**Comment by j9ac9k** [comment]
This LGTM, merging. 

@ntjess thanks so much for this.  This has been a feature I've wanted in the example app for so long, and on my machine, so often I grep a directory looking for an example of something....

Also thanks for my nit-pickiness through this 😆

**Comment by ntjess** [comment]
:wave: Happy to help! Also I greatly appreciate so many folks promptly pitching in... It greatly eased the troubleshooting process

**Pull request merged**